macOS: use file parent dir for `openTerminal` service cwd (#7286)

liby opened the pull request:
Fixes #7286

Previously, when using the "New Ghostty Window/Tab Here" macOS service on a file, the new terminal window/tab would incorrectly open in the user's home directory. This was because the service handler only expected directory paths.

This commit updates the service handler to check if the provided path is a file. If it is, the handler now uses the file's parent
directory as the working directory for the new Ghostty window or tab, aligning with user expectations. If the path is a directory, it's used directly as before.

liby committed:
macOS: use file parent dir for `openTerminal` service cwd (#7286)

jparise reviewed (commented):
I think is the right general approach, but I also think the implementation could be simplified to something like this:

```diff
diff --git i/macos/Sources/Features/Services/ServiceProvider.swift w/macos/Sources/Features/Services/ServiceProvider.swift
index bb95cb55..f0556160 100644
--- i/macos/Sources/Features/Services/ServiceProvider.swift
+++ w/macos/Sources/Features/Services/ServiceProvider.swift
@@ -47,14 +47,18 @@ class ServiceProvider: NSObject {
         let terminalManager = delegate.terminalManager
 
         for path in paths {
             // We only open in directories.
             var isDirectory = ObjCBool(true)
             guard FileManager.default.fileExists(atPath: path, isDirectory: &isDirectory) else { continue }
-            guard isDirectory.boolValue else { continue }
+
+            var workingDirectory = path
+            if !isDirectory.boolValue {
+                workingDirectory = (path as NSString).deletingLastPathComponent
+                guard FileManager.default.fileExists(atPath: workingDirectory, isDirectory: &isDirectory), isDirectory.boolValue else { continue }
+            }
 
             // Build our config
             var config = Ghostty.SurfaceConfiguration()
-            config.workingDirectory = path
+            config.workingDirectory = workingDirectory
 
             switch (target) {
             case .window:

```

liby committed:
macOS: simplify path handling in `openTerminal`

Aaron-212 commented:
I think I just found a simpler approach 

```diff
diff --git a/macos/Sources/Features/Services/ServiceProvider.swift b/macos/Sources/Features/Services/ServiceProvider.swift
index bb95cb55..5aceb8d5 100644
--- a/macos/Sources/Features/Services/ServiceProvider.swift
+++ b/macos/Sources/Features/Services/ServiceProvider.swift
@@ -36,25 +36,23 @@ class ServiceProvider: NSObject {
             error.pointee = Self.errorNoString
             return
         }
-        let filePaths = objs.map { $0.path }.compactMap { $0 }
+        let filePaths = objs.map { $0 as URL }

         openTerminal(filePaths, target: target)
     }

-    private func openTerminal(_ paths: [String], target: OpenTarget) {
+    private func openTerminal(_ urls: [URL], target: OpenTarget) {
         guard let delegateRaw = NSApp.delegate else { return }
         guard let delegate = delegateRaw as? AppDelegate else { return }
         let terminalManager = delegate.terminalManager

-        for path in paths {
+        for url in urls {
             // We only open in directories.
-            var isDirectory = ObjCBool(true)
-            guard FileManager.default.fileExists(atPath: path, isDirectory: &isDirectory) else { continue }
-            guard isDirectory.boolValue else { continue }
+            let cwd = url.hasDirectoryPath ? url : url.deletingLastPathComponent()

             // Build our config
             var config = Ghostty.SurfaceConfiguration()
-            config.workingDirectory = path
+            config.workingDirectory = cwd.path(percentEncoded: false)

             switch (target) {
             case .window:
```

tested against some files and folders and it *seems* to work fine

liby committed:
macOS: switch to using URL instead of String

liby commented:
Thanks for the pointer — the `URL`-based version is much nicer! I’ve adopted it and just added two small tweaks:

1. Wrap the CWD URLs in a `Set` so we don’t open the same directory multiple times when several files in it are selected.
2. Renamed a couple of variables for clarity (`urlObjects`, `uniqueCwds`, etc.).

Everything else is exactly your diff. Appreciate the help!

mitchellh reviewed (approved):
Awesome thanks everyone.
